### PR TITLE
Validate required fields of structs

### PR DESCRIPTION
**Pull request by kriskowal**
r @malandrew @ShanniLi

### CODE REVIEW

File: struct.js [hunk 1]
```diff
@@ -187,6 +187,29 @@ ThriftStruct.prototype.link = function link(spec) {
...
+    var self = this;
+
+    // Validate required fields
+    for (var index = 0; index < self.fields.length; index++) {
+        var field = self.fields[index];
+        if (!field.required || field.defaultValue != null) {
+            continue;
+        }
+        var value = struct && struct[field.name];
+        var available = value !== null && value !== undefined;
```
Comment (ShanniLi) [inline struct.js]: Could 'null' be a valid value?

File: struct.js [hunk 2]
```diff
@@ -381,6 +404,12 @@ StructRW.prototype.readFrom = function readFrom(buffer, offset) {
         self.spec.set(struct, field.name, result.value);
     }
 
+    // Validate required fields
+    var err = self.spec.validateStruct(struct);
```
Comment (ShanniLi) [inline struct.js]: Is there recursive, i.e., work on nested structs?

**Comment by ShanniLi** [comment]
Two questions. Otherwise, lgtm.

File: struct.js (hunk 1 above)
Comment (kriskowal) [inline struct.js]: null is not expressible except as property omission in Thrift.

File: struct.js (hunk 2 above)
Comment (kriskowal) [inline struct.js]: Each struct is validated upon read, including deeply nested structs.

**Pull request merged**